File: ciede-2000.pl
# This function written in Perl is not affiliated with the CIE (International Commission on Illumination),
# and is released into the public domain. It is provided "as is" without any warranty, express or implied.

use strict;
use warnings;
use Math::Trig qw(pi);

# Classic color accuracy can be ensured with this ΔE, ΔE2000 (ΔE00) implementation.
sub ciede_2000 {
	# Working with the CIEDE2000 color-difference formula.
	# k_l, k_c, k_h are parametric factors to be adjusted according to
	# different viewing parameters such as textures, backgrounds...
	my ($l_1, $a_1, $b_1, $l_2, $a_2, $b_2) = @_;
	my ($k_l, $k_c, $k_h) = (1.0, 1.0, 1.0);
	my $n = (sqrt($a_1 * $a_1 + $b_1 * $b_1) + sqrt($a_2 * $a_2 + $b_2 * $b_2)) * 0.5;
	$n = $n * $n * $n * $n * $n * $n * $n;
	# A factor involving chroma raised to the power of 7 designed to make
	# the influence of chroma on the total color difference more accurate.
	$n = 1.0 + 0.5 * (1.0 - sqrt($n / ($n + 6103515625.0)));
	my $c_1 = sqrt($a_1 * $a_1 * $n * $n + $b_1 * $b_1);
	my $c_2 = sqrt($a_2 * $a_2 * $n * $n + $b_2 * $b_2);
	# atan2 is preferred over atan because it accurately computes the angle of
	# a point (x, y) in all quadrants, handling the signs of both coordinates.
	my $h_1 = atan2($b_1, $a_1 * $n);
	my $h_2 = atan2($b_2, $a_2 * $n);
	$h_1 += 2.0 * pi if $h_1 < 0.0;
	$h_2 += 2.0 * pi if $h_2 < 0.0;
	$n = abs($h_2 - $h_1);
	# Cross-implementation consistent rounding.
	$n = pi if pi - 1E-14 < $n && $n < pi + 1E-14;
	# When the hue angles lie in different quadrants, the straightforward
	# average can produce a mean that incorrectly suggests a hue angle in
	# the wrong quadrant, the next lines handle this issue.
	my $h_m = 0.5 * $h_1 + 0.5 * $h_2;
	my $h_d = ($h_2 - $h_1) * 0.5;
	if (pi < $n) {
		$h_d -= pi if $h_d > 0.0;
		$h_d += pi if $h_d <= 0.0;
		$h_m += pi;
	}
	my $p = (36.0 * $h_m - 55.0 * pi);
	$n = ($c_1 + $c_2) * 0.5;
	$n = $n * $n * $n * $n * $n * $n * $n;
	# The hue rotation correction term is designed to account for the
	# non-linear behavior of hue differences in the blue region.
	my $r_t = -2.0 * sqrt($n / ($n + 6103515625.0))
			* sin(pi / 3.0 * exp($p * $p / (-25.0 * pi * pi)));
	$n = ($l_1 + $l_2) * 0.5;
	$n = ($n - 50.0) * ($n - 50.0);
	# Lightness.
	my $l = ($l_2 - $l_1) / ($k_l * (1.0 + 0.015 * $n / sqrt(20.0 + $n)));
	# These coefficients adjust the impact of different harmonic
	# components on the hue difference calculation.
	my $t = 1.0 + 0.24 * sin(2.0 * $h_m + pi / 2.0)
				+ 0.32 * sin(3.0 * $h_m + 8.0 * pi / 15.0)
				- 0.17 * sin($h_m + pi / 3.0)
				- 0.20 * sin(4.0 * $h_m + 3.0 * pi / 20.0);
	$n = $c_1 + $c_2;
	# Hue.
	my $h = 2.0 * sqrt($c_1 * $c_2) * sin($h_d) / ($k_h * (1.0 + 0.0075 * $n * $t));
	# Chroma.
	my $c = ($c_2 - $c_1) / ($k_c * (1.0 + 0.0225 * $n));
	# Returning the square root ensures that the result represents
	# the "true" geometric distance in the color space.
	return sqrt($l * $l + $h * $h + $c * $c + $c * $h * $r_t);
}

#
# More samples for the CIEDE2000 color difference formula implementation at https:#bit.ly/ciede2000-samples
#
# Testing lab(68.8, 31.7, 61.68) vs lab(68.8, 31.7, 61.68) expects ΔE2000 = ΔE00 = 0.0
# Testing lab(73.4489, 2.624, -80.948) vs lab(73.4489, 2.624, -81.0) expects ΔE2000 = ΔE00 = 0.01199644243
# Testing lab(88.8152, 126.103, -58.983) vs lab(88.8152, 125.91, -58.983) expects ΔE2000 = ΔE00 = 0.03209189129
# Testing lab(28.1, -24.1, -63.8) vs lab(28.1, -24.1, -64.064) expects ΔE2000 = ΔE00 = 0.05910096658
# Testing lab(35.0, -74.6503, -108.563) vs lab(35.1, -74.6503, -108.563) expects ΔE2000 = ΔE00 = 0.08231513909
# Testing lab(9.0, 4.4, 97.9904) vs lab(9.0, 4.4, 99.0) expects ΔE2000 = ΔE00 = 0.18703973501
# Testing lab(34.1, 15.34, -69.673) vs lab(34.72, 15.34, -69.673) expects ΔE2000 = ΔE00 = 0.50621159851
# Testing lab(30.33, 86.07, 110.0) vs lab(31.0, 86.07, 108.3) expects ΔE2000 = ΔE00 = 0.71295691566
# Testing lab(34.0, -13.723, -122.0462) vs lab(34.0, -10.9932, -122.0462) expects ΔE2000 = ΔE00 = 1.04546862433
# Testing lab(8.36, 123.13, 92.89) vs lab(8.36, 123.13, 89.0) expects ΔE2000 = ΔE00 = 1.19607699383
# Testing lab(46.61, -10.9, -97.151) vs lab(46.61, -6.687, -103.2) expects ΔE2000 = ΔE00 = 1.38274601092
# Testing lab(104.4427, -85.0, 100.0) vs lab(104.4427, -92.9713, 100.0) expects ΔE2000 = ΔE00 = 1.81795930601
# Testing lab(85.04, -51.503, 23.2) vs lab(85.04, -56.0, 28.216) expects ΔE2000 = ΔE00 = 2.06083109438
# Testing lab(42.0, -39.47, -31.49) vs lab(42.0, -39.47, -37.2552) expects ΔE2000 = ΔE00 = 2.32242738733
# Testing lab(56.529, -46.0, -57.73) vs lab(56.901, -39.0, -57.73) expects ΔE2000 = ΔE00 = 2.39806726715
# Testing lab(30.206, -50.4012, 6.1355) vs lab(30.206, -59.776, 6.1355) expects ΔE2000 = ΔE00 = 2.7240290859
# Testing lab(70.0044, 28.13, -13.0) vs lab(70.0044, 28.13, -8.0) expects ΔE2000 = ΔE00 = 2.89547379255
# Testing lab(13.1718, 92.99, -47.1726) vs lab(18.0, 92.99, -47.1726) expects ΔE2000 = ΔE00 = 3.19344984117
# Testing lab(10.12, -45.0558, 14.3) vs lab(15.09, -45.0558, 17.0) expects ΔE2000 = ΔE00 = 3.47496337913
# Testing lab(2.04, 39.885, 96.0001) vs lab(2.04, 32.11, 95.285) expects ΔE2000 = ΔE00 = 3.74387063768
# Testing lab(85.916, 6.0, 73.9) vs lab(85.916, -0.484, 73.9) expects ΔE2000 = ΔE00 = 3.87763601902
# Testing lab(63.0023, 20.0, -83.721) vs lab(66.8, 20.0, -93.06) expects ΔE2000 = ΔE00 = 4.31234276522
# Testing lab(45.9, -126.2, 58.56) vs lab(50.272, -126.2, 55.75) expects ΔE2000 = ΔE00 = 4.37977198118
# Testing lab(120.6, 67.595, 48.0) vs lab(127.2, 67.595, 56.58) expects ΔE2000 = ΔE00 = 4.82643694592
# Testing lab(113.0, -8.521, -38.0464) vs lab(112.7, -16.4175, -43.302) expects ΔE2000 = ΔE00 = 5.02928774761
# Testing lab(17.37, 114.9, -22.7699) vs lab(21.09, 116.0, -5.1136) expects ΔE2000 = ΔE00 = 5.70475201263
# Testing lab(82.6, -55.0, -30.7511) vs lab(80.0, -75.61, -47.075) expects ΔE2000 = ΔE00 = 6.32496865878
# Testing lab(100.39, -118.57, -54.0) vs lab(95.066, -104.3, -72.0) expects ΔE2000 = ΔE00 = 7.52227276077
# Testing lab(45.445, 84.628, 109.4) vs lab(42.0, 117.0, 121.3) expects ΔE2000 = ΔE00 = 8.28870814997
# Testing lab(93.7, -30.804, 40.3978) vs lab(109.87, -39.0, 46.9) expects ΔE2000 = ΔE00 = 9.63467721686
# Testing lab(123.0, 98.34, 39.4355) vs lab(114.2, 66.0, 14.41) expects ΔE2000 = ΔE00 = 10.82655645796
# Testing lab(81.4, 55.5, 63.9) vs lab(68.414, 37.2159, 54.0) expects ΔE2000 = ΔE00 = 11.50485227466
# Testing lab(19.975, 97.9434, -100.0) vs lab(6.481, 56.0, -71.21) expects ΔE2000 = ΔE00 = 12.5654122779
# Testing lab(20.871, 54.0, 32.0) vs lab(18.466, 67.5951, 8.9079) expects ΔE2000 = ΔE00 = 13.60173870607
# Testing lab(104.22, -68.673, 47.5) vs lab(82.0, -101.0, 59.4799) expects ΔE2000 = ΔE00 = 15.02268848663
# Testing lab(96.0393, -72.8, -0.14) vs lab(122.537, -92.33, 13.8) expects ΔE2000 = ΔE00 = 15.67150076127
# Testing lab(97.168, 18.83, -66.501) vs lab(124.0982, -8.09, -32.0) expects ΔE2000 = ΔE00 = 16.73938080381
# Testing lab(106.54, -106.0, 66.695) vs lab(79.0, -87.981, 50.944) expects ΔE2000 = ΔE00 = 17.29454469889
# Testing lab(33.651, -78.372, 70.689) vs lab(11.302, -38.76, 55.254) expects ΔE2000 = ΔE00 = 19.04568979077
# Testing lab(127.848, -83.1814, 60.0) vs lab(104.89, -89.0, 13.0) expects ΔE2000 = ΔE00 = 19.79352063552
# Testing lab(77.9, 119.1, -85.8043) vs lab(108.2998, 127.0, -127.3) expects ΔE2000 = ΔE00 = 20.49408851395
# Testing lab(80.0, 111.2, -49.1) vs lab(57.0, 108.8514, -4.4) expects ΔE2000 = ΔE00 = 21.97987691506
# Testing lab(116.792, 62.0, -6.0) vs lab(86.5, 56.016, -41.68) expects ΔE2000 = ΔE00 = 22.65807761353
# Testing lab(90.22, 86.516, 48.328) vs lab(117.6, 28.9, 31.9155) expects ΔE2000 = ΔE00 = 23.47669645731
# Testing lab(100.5443, -76.593, 98.008) vs lab(90.0, -89.13, 18.02) expects ΔE2000 = ΔE00 = 24.36773576669
# Testing lab(57.6726, -26.7, -126.3) vs lab(88.89, -35.7, -66.24) expects ΔE2000 = ΔE00 = 25.49940582851
# Testing lab(44.0, 2.0, 88.162) vs lab(123.8396, 31.0, 34.8891) expects ΔE2000 = ΔE00 = 60.74253483449
# Testing lab(38.9364, -18.5, -64.7194) vs lab(122.3013, -50.181, 54.0) expects ΔE2000 = ΔE00 = 83.65902437262
# Testing lab(79.601, -122.0, -40.0) vs lab(15.0457, 77.348, -16.6) expects ΔE2000 = ΔE00 = 123.35005156744
# Testing lab(12.14, 126.95, -63.838) vs lab(99.0, -67.68, 66.3) expects ΔE2000 = ΔE00 = 135.41043955769

File: ciede-2000.php
<?php

// This function written in PHP is not affiliated with the CIE (International Commission on Illumination),
// and is released into the public domain. It is provided "as is" without any warranty, express or implied.

// Classic color accuracy can be ensured with this ΔE, ΔE2000 (ΔE00) implementation.
function ciede_2000($L1, $a1, $b1, $L2, $a2, $b2) {
	// Working with the CIEDE2000 color-difference formula.
	// kL, kC, kH are parametric factors to be adjusted according to
	// different viewing parameters such as textures, backgrounds...
	$kL = $kC = $kH = 1.0;
	$n = (hypot($a1, $b1) + hypot($a2, $b2)) * 0.5;
	$n = $n * $n * $n * $n * $n * $n * $n;
	// A factor involving chroma raised to the power of 7 designed to make
	// the influence of chroma on the total color difference more accurate.
	$n = 1.0 + 0.5 * (1.0 - sqrt($n / ($n + 6103515625.0)));
	// hypot calculates the Euclidean distance while avoiding overflow/underflow.
	$c_1 = hypot($a1 * $n, $b1);
	$c_2 = hypot($a2 * $n, $b2);
	// atan2 is preferred over atan because it accurately computes the angle of
	// a point (x, y) in all quadrants, handling the signs of both coordinates.
	$h_1 = atan2($b1, $a1 * $n);
	$h_2 = atan2($b2, $a2 * $n);
	$h_1 += 2.0 * M_PI * ($h_1 < 0.0);
	$h_2 += 2.0 * M_PI * ($h_2 < 0.0);
	// When the hue angles lie in different quadrants, the straightforward
	// average can produce a mean that incorrectly suggests a hue angle in
	// the wrong quadrant, the next line handle this issue.
	$h_tmp = ($h_1 + $h_2) * 0.5 + M_PI * (M_PI < abs($h_1 - $h_2));
	$n = ($c_1 + $c_2) * 0.5;
	$n = $n * $n * $n * $n * $n * $n * $n;
	$p = (36.0 * $h_tmp - 55.0 * M_PI);
	// The hue rotation correction term is designed to account for the
	// non-linear behavior of hue differences in the blue region.
	$r_t = -2.0	* sqrt($n / ($n + 6103515625.0))
			* sin(M_PI / 3.0 * exp($p * $p / (-25.0 * M_PI * M_PI)));
	$n = ($L1 + $L2) * 0.5;
	$n = ($n - 50.0) * ($n - 50.0);
	// Lightness
	$l = ($L2 - $L1) / ($kL * (1.0 + 0.015 * $n / sqrt(20.0 + $n)));
	// These coefficients adjust the impact of different harmonic
	// components on the hue difference calculation.
	$t = 1.0	+ 0.24 * sin(2.0 * $h_tmp + M_PI_2)
			+ 0.32 * sin(3.0 * $h_tmp + 8.0 * M_PI / 15.0)
			- 0.17 * sin($h_tmp + M_PI / 3.0)
			- 0.20 * sin(4.0 * $h_tmp + 3.0 * M_PI_2 / 10.0);
	$n = $c_1 + $c_2;
	$h_tmp = ($h_2 - $h_1) * 0.5;
	$h_tmp += M_PI * ($h_tmp < -M_PI_2);
	$h_tmp -= M_PI * (M_PI_2 < $h_tmp);
	// Hue
	$h = 2.0 * sqrt($c_1 * $c_2) * sin($h_tmp) / ($kH * (1.0 + 0.0075 * $n * $t));
	// Chroma
	$c = ($c_2 - $c_1) / ($kC * (1 + 0.0225 * $n));
	// Returning the square root ensures that the result represents
	// the "true" geometric distance in the color space.
	return sqrt($l * $l + $h * $h + $c * $c + $c * $h * $r_t);
}

//
// More samples for the CIEDE2000 color difference formula implementation at https://bit.ly/ciede2000-samples
//
// Testing lab(76.844, -14.388, 40.4585) vs lab(76.844, -14.388, 40.4585) expects ΔE2000 = ΔE00 = 0.0
// Testing lab(54.844, -119.8, 111.0) vs lab(54.844, -119.8, 111.123) expects ΔE2000 = ΔE00 = 0.02313555897
// Testing lab(27.0, -61.85, 60.535) vs lab(27.0, -61.85, 60.371) expects ΔE2000 = ΔE00 = 0.04876786479
// Testing lab(5.5, 111.6147, -108.0) vs lab(5.5, 111.6147, -107.789) expects ΔE2000 = ΔE00 = 0.05440969008
// Testing lab(47.0, -93.091, -89.2) vs lab(47.0, -92.711, -89.2) expects ΔE2000 = ΔE00 = 0.07958280852
// Testing lab(61.8, 56.71, 41.0) vs lab(61.8, 56.71, 40.71) expects ΔE2000 = ΔE00 = 0.13823262029
// Testing lab(119.0, 29.0, 85.61) vs lab(120.0, 29.0, 85.61) expects ΔE2000 = ΔE00 = 0.4901123735
// Testing lab(89.0, 77.0, 12.39) vs lab(89.9, 75.9, 12.39) expects ΔE2000 = ΔE00 = 0.62086082535
// Testing lab(117.264, 107.3, -92.0) vs lab(117.264, 111.1, -92.0) expects ΔE2000 = ΔE00 = 0.85920944045
// Testing lab(47.09, -72.753, 70.8246) vs lab(47.09, -72.753, 75.1226) expects ΔE2000 = ΔE00 = 1.12797574463
// Testing lab(35.8213, -5.8, -113.18) vs lab(35.8213, -9.2132, -113.18) expects ΔE2000 = ΔE00 = 1.42607780642
// Testing lab(125.0, 121.0, 69.811) vs lab(125.0, 118.8, 74.0) expects ΔE2000 = ΔE00 = 1.76960674261
// Testing lab(19.45, 95.3, 110.5) vs lab(22.28, 95.3, 108.834) expects ΔE2000 = ΔE00 = 2.03359458921
// Testing lab(15.2747, -88.0, -46.0) vs lab(15.2747, -88.0, -39.623) expects ΔE2000 = ΔE00 = 2.21543785112
// Testing lab(126.11, -48.51, 80.4) vs lab(126.11, -41.47, 80.4) expects ΔE2000 = ΔE00 = 2.56244534184
// Testing lab(91.6907, 77.0, -59.0) vs lab(91.6907, 77.0, -67.6) expects ΔE2000 = ΔE00 = 2.84932248426
// Testing lab(95.42, 38.0, -1.2629) vs lab(98.3671, 43.502, -3.824) expects ΔE2000 = ΔE00 = 2.90347144498
// Testing lab(45.85, -38.0, 119.7) vs lab(49.1, -38.0, 119.7) expects ΔE2000 = ΔE00 = 3.19058676452
// Testing lab(36.276, 45.4271, -22.9308) vs lab(36.276, 41.0, -28.0) expects ΔE2000 = ΔE00 = 3.41413009713
// Testing lab(103.1, -78.5, -117.7) vs lab(110.0, -78.5, -116.0) expects ΔE2000 = ΔE00 = 3.74894217108
// Testing lab(125.79, 26.8, 67.0) vs lab(133.5308, 26.8, 61.9488) expects ΔE2000 = ΔE00 = 3.89352301119
// Testing lab(5.52, 34.24, 64.27) vs lab(7.189, 42.0, 64.27) expects ΔE2000 = ΔE00 = 4.1744218339
// Testing lab(127.413, 111.3596, -100.5197) vs lab(127.413, 101.4, -109.3501) expects ΔE2000 = ΔE00 = 4.53539868441
// Testing lab(27.0, -15.4, -72.0) vs lab(30.0, -24.707, -69.1719) expects ΔE2000 = ΔE00 = 4.66523571065
// Testing lab(41.7135, 93.95, 8.236) vs lab(46.0, 109.425, 13.0) expects ΔE2000 = ΔE00 = 5.01544767819
// Testing lab(112.6, -112.85, 16.96) vs lab(124.321, -110.4, 16.63) expects ΔE2000 = ΔE00 = 5.80321420993
// Testing lab(84.0, -104.292, 81.7496) vs lab(76.7, -112.0, 105.55) expects ΔE2000 = ΔE00 = 6.66326863289
// Testing lab(98.0, 69.11, -98.46) vs lab(110.0, 84.241, -121.0) expects ΔE2000 = ΔE00 = 7.71613301575
// Testing lab(126.0, 47.2, 65.542) vs lab(124.0, 30.2, 62.5) expects ΔE2000 = ΔE00 = 8.15905991594
// Testing lab(110.8, -100.022, -3.5936) vs lab(116.0, -120.25, 22.0) expects ΔE2000 = ΔE00 = 9.94593509197
// Testing lab(117.1, 28.0, 54.58) vs lab(111.94, 54.67, 93.5) expects ΔE2000 = ΔE00 = 10.40882182343
// Testing lab(92.251, -124.925, -122.3383) vs lab(84.4, -108.29, -65.8848) expects ΔE2000 = ΔE00 = 11.41424538448
// Testing lab(86.343, 77.803, -50.536) vs lab(107.406, 74.2, -43.603) expects ΔE2000 = ΔE00 = 12.55528510533
// Testing lab(94.426, -16.8, -75.0) vs lab(90.0, -53.0, -105.8675) expects ΔE2000 = ΔE00 = 13.1700969699
// Testing lab(31.049, -91.4, 57.5) vs lab(30.0, -74.048, 106.5) expects ΔE2000 = ΔE00 = 14.30590820471
// Testing lab(115.0, 2.4, -26.7) vs lab(114.96, 67.0, -111.605) expects ΔE2000 = ΔE00 = 16.08348293502
// Testing lab(67.4935, 67.22, -82.518) vs lab(77.0, 98.577, -57.7) expects ΔE2000 = ΔE00 = 16.57071380869
// Testing lab(1.2039, -112.677, 89.5092) vs lab(24.0, -76.0, 43.63) expects ΔE2000 = ΔE00 = 17.91366640754
// Testing lab(101.0, 6.97, -30.2091) vs lab(79.657, 30.257, -86.0) expects ΔE2000 = ΔE00 = 18.65098697254
// Testing lab(74.125, -99.0, -60.2546) vs lab(67.8461, -51.0, -110.89) expects ΔE2000 = ΔE00 = 19.12450228953
// Testing lab(88.25, -97.7, -102.0) vs lab(125.935, -103.42, -104.08) expects ΔE2000 = ΔE00 = 20.35192502517
// Testing lab(124.2, -56.8, -87.2) vs lab(89.1, -115.91, -112.0) expects ΔE2000 = ΔE00 = 21.99317447199
// Testing lab(32.0, -59.7704, 60.0) vs lab(2.0, -41.4871, 75.3408) expects ΔE2000 = ΔE00 = 22.17460912079
// Testing lab(103.8, 53.0, 101.8) vs lab(77.4476, 12.97, 51.913) expects ΔE2000 = ΔE00 = 23.3352467278
// Testing lab(85.98, 126.3, 66.5301) vs lab(57.0, 118.482, 34.153) expects ΔE2000 = ΔE00 = 24.15637515461
// Testing lab(61.3625, -77.54, -84.8437) vs lab(100.242, -122.26, -101.2852) expects ΔE2000 = ΔE00 = 27.77670386286
// Testing lab(92.9, -115.4111, 10.3015) vs lab(71.472, 42.4, -114.0087) expects ΔE2000 = ΔE00 = 54.79818740897
// Testing lab(124.279, 105.362, -66.6) vs lab(42.0, 31.85, 93.0) expects ΔE2000 = ΔE00 = 86.44280474288
// Testing lab(53.59, 118.2, -52.575) vs lab(8.802, -53.09, 69.0) expects ΔE2000 = ΔE00 = 106.74910649823
// Testing lab(119.294, 88.7558, 116.97) vs lab(2.7646, -123.46, 34.6087) expects ΔE2000 = ΔE00 = 129.98829153146

